stretch_core/stretch_core/stretch_driver.py

```diff
@@ -572,7 +586,8 @@ def ros_setup(self):
         self.magnetometer_mobile_base_pub = self.create_publisher(MagneticField, 'magnetometer_mobile_base', 1)
         self.imu_wrist_pub = self.create_publisher(Imu, 'imu_wrist', 1)
 
-        self.create_subscription(Twist, "cmd_vel", self.set_mobile_base_velocity_callback, 1)
+        self.group = MutuallyExclusiveCallbackGroup()
```

Comment:
What does `MutuallyExclusiveCallbackGroup` do?

Reply:
This might be slightly difficult to explain. In multithreaded execution, callback groups allow separating callbacks in different groups. This is necessary when you wouldn't want two callbacks that are reading/writing from/to shared memory to run parallelly to prevent deadlocks and race conditions.  The MutuallyExclusiveCallbackGroup ensures that only one callback has access to shared memory at a given time. 

In our case, it was necessary to separately execute set_mobile_base_velocity_callback() and command_mobile_base_velocity_and_publish_state() because they both access self.linear_velocity_mps and
self.angular_velocity_radps. I believe a potential deadlock was resulting in the choppy navigation issue. Introducing MutuallyExclusiveCallbackGroup resolved this.

I'd recommend reading [Deadlocks in rclpy and how to prevent them with use of callback groups](https://karelics.fi/deadlocks-in-rclpy/) to fill the gaps in my explanation.